File: OsmAnd/src/net/osmand/plus/OnNavigationServiceAlarmReceiver.java
package net.osmand.plus;

import android.annotation.SuppressLint;
import android.app.AlarmManager;
import android.app.PendingIntent;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.PowerManager.WakeLock;
import android.os.SystemClock;

public class OnNavigationServiceAlarmReceiver extends BroadcastReceiver {
	@SuppressLint("WakelockTimeout")
	@Override
	public void onReceive(Context context, Intent intent) {
		final WakeLock lock = NavigationService.getLock(context);
		final NavigationService service = ((OsmandApplication) context.getApplicationContext()).getNavigationService();

		// do not do nothing
		if (service == null) {
			return;
		} else if (lock.isHeld()) {
			rescheduleAlarm(context, service);
			return;
		}

		lock.acquire();
		rescheduleAlarm(context, service);
		service.onWakeUp();
	}

	private void rescheduleAlarm(Context context, NavigationService service) {
		//Unless setRepeating was used, manually re-schedule service to the next measurement point in the future
		int serviceOffInterval = service.getServiceOffInterval();
		if (Build.VERSION.SDK_INT >= 19 && serviceOffInterval > 0) {
			// Avoid drift
			while ((service.getNextManualWakeup() - SystemClock.elapsedRealtime()) < 0) {
				service.setNextManualWakeup(service.getNextManualWakeup() + serviceOffInterval);
			}
			AlarmManager alarmManager = (AlarmManager) service.getSystemService(Context.ALARM_SERVICE);
			PendingIntent pendingIntent = PendingIntent.getBroadcast(context, 1, new Intent(context, OnNavigationServiceAlarmReceiver.class), PendingIntent.FLAG_UPDATE_CURRENT);
			if (Build.VERSION.SDK_INT >= 23) {
				alarmManager.setExactAndAllowWhileIdle(AlarmManager.ELAPSED_REALTIME_WAKEUP, service.getNextManualWakeup(), pendingIntent);
			} else {
				alarmManager.setExact(AlarmManager.ELAPSED_REALTIME_WAKEUP, service.getNextManualWakeup(), pendingIntent);
			}
		}
	}
}

File: OsmAnd-telegram/src/net/osmand/telegram/OnNavigationServiceAlarmReceiver.kt
package net.osmand.telegram

import android.annotation.SuppressLint
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.location.LocationManager
import net.osmand.telegram.TelegramApplication
import net.osmand.telegram.TelegramService
import net.osmand.telegram.utils.AndroidUtils

class OnNavigationServiceAlarmReceiver : BroadcastReceiver() {
	@SuppressLint("MissingPermission")
	override fun onReceive(context: Context, intent: Intent) {
		val lock = TelegramService.getLock(context)
		val app = context.applicationContext as TelegramApplication
		val service = app.telegramService
		// do not do nothing
		if (lock.isHeld || service == null) {
			return
		}
		lock.acquire(10 * 60 * 1000L /*10 minutes*/)

		// request location updates
		val locationManager = service.getSystemService(Context.LOCATION_SERVICE) as LocationManager
		try {
			if (AndroidUtils.isLocationPermissionAvailable(app)) {
				locationManager.requestLocationUpdates(service.serviceOffProvider, 0, 0f, service)
				val handler = service.handler
				if (service.serviceOffInterval > service.serviceError && handler != null) {
					handler.postDelayed({
						// if lock is not anymore held
						if (lock.isHeld) {
							lock.release()
							locationManager.removeUpdates(service)
						}
					}, service.serviceError)
				}
			}
		} catch (e: RuntimeException) {
			e.printStackTrace()
		}
	}
}